R/group.R

```diff
@@ -0,0 +1,71 @@
+#' @title Grouping of numeric values by similarity
+#'
+#' @description
+#'
+#' The `group` function groups numeric values by first ordering and then putting
+#' all values into the same group if their difference is smaller defined by
+#' parameters `tolerance` (a constant value) and `ppm` (a value-specific
+#' relative value expressed in parts-per-million).
+#'
+#' @note
+#'
+#' Since grouping is performed on pairwise differences between consecutive
+#' values (after ordering `x`), the difference between the smallest and largest
+#' value in a group can be larger than `tolerance` and `ppm`.
+#'
+#' @param x increasingly ordered `numeric` with the values to be grouped.
+#'
+#' @param tolerance `numeric(1)` with the maximal accepted difference between
+#'     values in `x` to be grouped into the same entity.
+#'
+#' @param ppm `numeric(1)` defining a value-dependent maximal accepted
+#'     difference between values in `x` expressed in parts-per-million.
+#'
+#' @return `integer` of length equal to `x` with the groups.
+#'
+#' @author Johannes Rainer
+#'
+#' @rdname group
+#'
+#' @export group
+#'
+#' @examples
+#'
+#' ## Define a (sorted) numeric vector
+#' x <- c(34, 35, 35, 35 + ppm(35, 10), 56, 56.05, 56.1)
+#'
+#' ## With `ppm = 0` and `tolerance = 0` only identical values are grouped
+#' group(x)
+#'
+#' ## With `tolerance = 0.05`
+#' group(x, tolerance = 0.05)
+#'
+#' ## Also values 56, 56.05 and 56.1 were grouped into a single group,
+#' ## although the difference between the smallest 56 and largest value in
+#' ## this group (56.1) is 0.1. The (pairwise) difference between the ordered
+#' ## values is however 0.05.
+#'
+#' ## With ppm
+#' group(x, ppm = 10)
+#'
+#' ## Same on an unsorted vector
+#' x <- c(65, 34, 65.1, 35, 66, 65.2)
+#' group(x, tolerance = 0.1)
+#'
+#' ## Values 65, 65.1 and 65.2 have been grouped into the same group.
+group <- function(x, tolerance = 0, ppm = 0) {
+    if (is.unsorted(x)) {
+        idx <- order(x)
+        x <- x[idx]
+    } else idx <- integer()
+    xdiff <- diff(x)
+    tolerance <- tolerance + sqrt(.Machine$double.eps)
+    if (ppm > 0) {
+        res <- cumsum(c(0L, xdiff >= (tolerance +
+                                      ppm(x[-length(x)], ppm)))) + 1L
+    } else
+        res <- cumsum(c(0L, xdiff >= tolerance)) + 1L
+    if (length(idx))
+        res[order(idx)]
```

Comment:
The additional `order` call slows things down, why not use the index directly? I would rewrite the function as follows:
```r
group1 <- function(x, tolerance = 0, ppm = 0) {
    if (is.unsorted(x)) {
        idx <- order(x)
        x <- x[idx]
    } else
        idx <- integer()

    tolerance <- tolerance + sqrt(.Machine$double.eps)

    if (ppm > 0)
        tolerance <- tolerance + ppm(x[-length(x)], ppm)

    x <- cumsum(c(1L, diff(x) >= tolerance))
    x[idx] <- x
    x
}

set.seed(2019)
n <- 1e5
mz <- sample(n, n, replace = TRUE) + rnorm(n)
smz <- sort(mz)

microbenchmark(group(mz, ppm = 20), group1(mz, ppm = 20), check = "equal")
# Unit: milliseconds
#                  expr       min        lq     mean    median       uq      max
#   group(mz, ppm = 20) 10.436601 10.807413 11.59656 11.210812 12.14606 15.35732
#  group1(mz, ppm = 20)  8.423567  8.750131 10.16782  9.096824 10.10651 73.36924
#  neval
#    100
#    100
microbenchmark(group(smz, ppm = 20), group1(smz, ppm = 20), check = "equal")
# Unit: milliseconds
#                   expr      min       lq     mean   median       uq      max
#   group(smz, ppm = 20) 3.267696 3.421522 3.852874 3.536960 4.240001 7.588491
#  group1(smz, ppm = 20) 3.094932 3.245265 3.685689 3.321149 3.517340 7.728484
#  neval
#    100
#    100
microbenchmark(group(smz), group1(smz), check = "equal")
# Unit: milliseconds
#         expr      min       lq     mean   median       uq      max neval
#   group(smz) 2.398114 2.482195 2.897100 2.554832 2.785286 5.241478   100
#  group1(smz) 2.256011 2.304851 2.621905 2.385354 2.616277 5.361702   100
```